src/core/pile.sol

```diff
@@ -17,150 +17,203 @@ pragma solidity >=0.4.24;
-        return (debt, loans_[loan].balance, loans_[loan].rate);
-    }
+    // loan => pie
+    // pie = debt/chi
+    mapping (uint => uint) public pie;
+    // loan => rate
+    mapping (uint => uint) public group;
 
-    function Debt() public view returns (uint debt) {
-        return debtRegister.totalDebt();
-    }
+    uint public total;
 
-    uint public Balance;
-    address public lender;
-
-    constructor(address tkn_, address title_, address debtRegister_) TitleOwned(title_) public {
+    constructor() public {
         wards[msg.sender] = 1;
-        tkn = TokenLike(tkn_);
-        debtRegister = DebtRegisterLike(debtRegister_);
+        rates[0].chi      = ONE;
+        rates[0].speed    = ONE;
+    }
+
+    // --- Public Debt Methods  ---
+    // increase the loan's debt by wad
+    function incDebt(uint loan, uint wad) public auth note {
+        uint rate = group[loan];
+        require(now <= rates[rate].rho);
+        uint delta = toPie(rates[rate].chi, wad);
+        pie[loan] = add(pie[loan], delta);
+        rates[rate].pie = add(rates[rate].pie, delta);
+        total = add(total, wad);
+    }
+
+    // decrease the loan's debt by wad
+    function decDebt(uint loan, uint wad) public auth note {
+        uint rate = group[loan];
+        require(now <= rates[rate].rho);
+        uint delta = toPie(rates[rate].chi, wad);
+        pie[loan] = sub(pie[loan], delta);
+        rates[rate].pie = sub(rates[rate].pie, delta);
+        total = sub(total, wad);
+    }
+
+    function debt(uint loan) public view returns (uint) {
+        uint rate = group[loan];
+        uint chi = rates[rate].chi;
+        if (now >= rates[rate].rho) {
+            (chi,) = compound(rate);
+        }
+        return fromPie(chi, pie[loan]);
+    }
+
+    function rateDebt(uint rate) public view returns (uint) {
+        uint chi = rates[rate].chi;
+        uint pie = rates[rate].pie;
+        
+        if (now >= rates[rate].rho) {
+            (chi,) = compound(rate);
+        } 
+        return fromPie(chi, pie);
+    } 
+
+    // --- Interest Rate Group Implementation ---
+
+    // set rate group for a loan
+    function setRate(uint loan, uint rate) public auth {
+        require(pie[loan] == 0, "non-zero-debt");
+        group[loan] = rate;
+    }
+
+    // change rate group for a loan
+    function changeRate(uint loan, uint newRate) public auth {
+        uint currentRate = group[loan];
+        drip(currentRate);
+        drip(newRate);
+        uint pie_ = pie[loan];
+        uint debt = fromPie(rates[currentRate].chi, pie_);
+        rates[currentRate].pie = sub(rates[currentRate].pie, pie_);
+        pie[loan] = toPie(rates[newRate].chi, debt);
+        rates[newRate].pie = add(rates[newRate].pie, pie[loan]);
+        group[loan] = newRate;
+    }
+
+    // set/change the interest rate of a rate category
+    function file(uint rate, uint speed_) public auth {
+        require(speed_ != 0);
```

Comment:
we need `rates[rate].chi == 0` here because changing the speed already in use rate group would work differently.

It would just update the `speed` and not re-init the `chi`

Reply:
agree. See comment above:

Currently when we changing the speed of a rate group we just resetting chi to 1, which leads to issue when the group already has debt.

Reply:
and we also need to add a method that safely updates the group speed